### PR TITLE
Rationalise serialisation abstractions

### LINKED ISSUE #92: Unify event deserialization for persistence and projections
Currently, projections have a different abstraction to persistence for deserialising events. Bringing these together will help us to reuse builder abstractions across the write and read sides.

### PR DESCRIPTION
**Pull request by daniel-smith**
Improve event serialisation abstractions. This allows use to reuse deserialisation logic on both the persistence and projection side - ultimately allowing us to share some builders, e.g. `UsePostgresStreamStore` is now shared.

Essentially, it means that the event repository and event dispatchers can reuse the same deserialisation abstractions.

One important change is that we now throw `EventDeserializeException` from `LoadEventsAsync` on the event repository. We were not using the on-error callback anywhere - which meant that deserialisation errors were essentially swallowed. It also meant that the raw data type needed to be exposed on the interface. Going forward, we need to consider if we need a better strategy to handle event load errors like this.

Also reshuffled a few things and deleted old code.

Closes #92

### CODE REVIEW

File: src/DomainBlocks.Core/Metadata/MetadataKeys.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+namespace DomainBlocks.Core.Metadata;
+
+public static class MetadataKeys
+{
+    public const string LoggedInUserName = "LoggedInUserName";
+    public const string MachineName = "MachineName";
+    public const string ServiceName = "ServiceName";
+    public const string UtcDateTime = "UtcDataTime";
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Core/Metadata/MetadataKeys.cs:8]: Typo here (which was already there before) 

UtcDataTime -> UtcDateTime

File: src/DomainBlocks.Core/Serialization/EventConverter.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        _eventAdapter = eventAdapter ?? throw new ArgumentNullException(nameof(eventAdapter));
+        _metadataContext = metadataContext ?? EventMetadataContext.CreateEmpty();
+    }
+
+    public ValueTask<object> DeserializeEvent(
+        TReadEvent readEvent,
+        Type eventTypeOverride = null,
+        CancellationToken cancellationToken = default)
+    {
+        var eventName = _eventAdapter.GetEventName(readEvent);
```
Comment (carlosrfernandez) [inline src/DomainBlocks.Core/Serialization/EventConverter.cs:41]: It feels a bit odd here that we get the event name from the IEventAdapter, then the type from the IEventNameMap as opposed to the "SerializeToWriteEvent" where we get the name from IEventNameMap and use the IEventAdapter to serialize. 

Maybe  we can have a look offline, as I might be missing some additional context here

**Review by carlosrfernandez — DISMISSED** [review]
Just some minor nits

File: src/DomainBlocks.Core/Metadata/MetadataKeys.cs (hunk 1 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/Metadata/MetadataKeys.cs:8]: Right. Yeah. This stuff isn't really used anywhere at the moment, but will fix.

File: src/DomainBlocks.Core/Serialization/EventConverter.cs (hunk 2 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/Serialization/EventConverter.cs:41]: The ideas is that the EventConverter is responsible for converting a "raw" event to an instance of a CRL type. To do that, it needs to:

1. Know the name of the event
2. Map that name to a CRL type
3. Deserialise the event

The adapter helps to hide away how to get the event name, and how to get the underlying raw data from `TEvent` to (de)serialise. For example, take a look at `EventStoreEventAdapter`. This way, the EventConverter doesn't care what `TEvent` is - it just passes it on to the adapter, which "adapts" the raw event to the serialiser. Make sense?

File: src/DomainBlocks.Core/Metadata/MetadataKeys.cs (hunk 1 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/Metadata/MetadataKeys.cs]: Fixed

File: src/DomainBlocks.Core/Serialization/EventConverter.cs (hunk 2 above)
Comment (daniel-smith) [reply src/DomainBlocks.Core/Serialization/EventConverter.cs:41]: ...and the reason we get the name from IEventNameMap in `SerializeToWriteEvent` is that it's the only place that knows how to map the CLR type to the string name.

For the other way around - deserialising - we can only get the name of the event from the raw event itself

**Review by carlosrfernandez — APPROVED** [review]

**Pull request merged**